File: agent-sdk/python/first_agent_with_error_handling.py
"""
First Agent with Error Handling

Demonstrates proper error handling, logging, and graceful failure
using the Claude Agent SDK.

Usage:
    uv run agent-sdk/python/first_agent_with_error_handling.py
"""

import asyncio
import logging
import sys

from dotenv import load_dotenv

from claude_agent_sdk import (
    AssistantMessage,
    CLIJSONDecodeError,
    CLINotFoundError,
    ClaudeAgentOptions,
    ClaudeSDKError,
    ProcessError,
    ResultMessage,
    TextBlock,
    query,
)

load_dotenv()

if sys.platform == "win32":
    asyncio.set_event_loop_policy(asyncio.WindowsSelectorEventLoopPolicy())

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s [%(levelname)s] %(name)s: %(message)s",
)
logger = logging.getLogger(__name__)


async def main():
    logger.info("Starting agent...")

    try:
        options = ClaudeAgentOptions(
            permission_mode="bypassPermissions",
            model="claude-sonnet-4-5",
            max_turns=5,
        )

        async for message in query(
            prompt="Read src/lib/actions.ts and explain what each server action does.",
            options=options,
        ):
            if isinstance(message, AssistantMessage):
                for block in message.content:
                    if isinstance(block, TextBlock):
                        print(block.text)
            elif isinstance(message, ResultMessage):
                logger.info(
                    "Agent completed. Cost: $%.4f, Duration: %dms",
                    message.cost_usd,
                    message.duration_ms,
                )

    except CLINotFoundError:
        logger.error(
            "Claude CLI not found. Ensure it is installed and on your PATH."
        )
        sys.exit(1)
    except ProcessError as e:
        logger.error("Process error: %s", e)
        sys.exit(1)
    except CLIJSONDecodeError as e:
        logger.error("Failed to decode CLI JSON output: %s", e)
        sys.exit(1)
    except ClaudeSDKError as e:
        logger.error("SDK error: %s", e)
        sys.exit(1)


if __name__ == "__main__":
    asyncio.run(main())

File: agent-sdk/typescript/src/first-agent-with-error-handling.ts
/**
 * First Agent with Error Handling (TypeScript)
 *
 * Demonstrates proper error handling and logging using the
 * Claude Agent SDK.
 *
 * Usage:
 *   npx tsx src/first-agent-with-error-handling.ts
 */

import "dotenv/config";
import {
  query,
  type Options,
  type SDKAssistantMessage,
  type SDKResultMessage,
} from "@anthropic-ai/claude-agent-sdk";

async function main() {
  console.log("[INFO] Starting agent...");

  try {
    const options: Options = {
      permissionMode: "bypassPermissions",
      model: "claude-sonnet-4-5",
      maxTurns: 5,
    };

    const response = query({
      prompt:
        "Read src/lib/actions.ts and explain what each server action does.",
      options,
    });

    for await (const message of response) {
      if (message.type === "assistant") {
        for (const block of message.message.content) {
          if (block.type === "text") {
            console.log(block.text);
          }
        }
      } else if (message.type === "result") {
        console.log(
          `[INFO] Agent completed. Cost: $${message.total_cost_usd.toFixed(4)}, Duration: ${message.duration_ms}ms`
        );
        console.log(`[INFO] Session ID: ${message.session_id}`);
      }
    }
  } catch (error) {
    if (error instanceof Error) {
      console.error(`[ERROR] ${error.name}: ${error.message}`);
    } else {
      console.error("[ERROR] Unknown error:", error);
    }
    process.exit(1);
  }
}

main();
